common/solver/idr_kernels.hpp.inc

```diff
@@ -0,0 +1,360 @@
+    ValueType *__restrict__ g_values, size_type g_stride,
+    const stopping_status *__restrict__ stop_status)
+{
+    const auto tidx = thread::get_thread_id_flat();
+    const auto row = tidx / g_k_stride;
+    const auto rhs = tidx % nrhs;
+
+    if (row >= size || rhs >= nrhs) {
+        return;
+    }
+
+    if (!stop_status[rhs].has_stopped()) {
+        g_values[row * g_stride + k * nrhs + rhs] =
+            g_k_values[row * g_k_stride + rhs];
+    }
+}
+
+
+template <typename ValueType>
+__global__ __launch_bounds__(default_block_size) void update_x_r_and_f_kernel(
+    size_type k, size_type size, size_type subspace_dim, size_type nrhs,
+    const ValueType *__restrict__ m_values, size_type m_stride,
+    const ValueType *__restrict__ g_values, size_type g_stride,
+    const ValueType *__restrict__ u_values, size_type u_stride,
+    ValueType *__restrict__ f_values, size_type f_stride,
+    ValueType *__restrict__ r_values, size_type r_stride,
+    ValueType *__restrict__ x_values, size_type x_stride,
+    const stopping_status *__restrict__ stop_status)
+{
+    const auto global_id = thread::get_thread_id_flat();
+    const auto row = global_id / x_stride;
+    const auto col = global_id % x_stride;
+
+    if (row >= size || col >= nrhs) {
+        return;
+    }
+
+    if (!stop_status[col].has_stopped()) {
+        const auto beta = f_values[k * f_stride + col] /
+                          m_values[k * m_stride + k * nrhs + col];
+        r_values[row * r_stride + col] -=
+            beta * g_values[row * g_stride + k * nrhs + col];
+        x_values[row * x_stride + col] +=
+            beta * u_values[row * u_stride + k * nrhs + col];
+
+        if (k < row && k + 1 < subspace_dim && row < subspace_dim) {
+            f_values[row * f_stride + col] -=
+                beta * m_values[row * m_stride + k * nrhs + col];
+        }
+    }
+}
+
+
+template <typename ValueType>
+__global__ __launch_bounds__(config::warp_size) void set_f_zeros(
+    size_type k, size_type nrhs, ValueType *__restrict__ f_values,
+    size_type f_stride)
+{
+    const auto tidx = thread::get_thread_id_flat();
+    if (tidx < nrhs) {
+        f_values[k * f_stride + tidx] = zero<ValueType>();
+    }
+}
+
+
+template <typename ValueType>
+__global__ __launch_bounds__(config::warp_size) void compute_omega_kernel(
+    size_type nrhs, const remove_complex<ValueType> kappa,
+    const ValueType *__restrict__ tht,
+    const remove_complex<ValueType> *__restrict__ residual_norm,
+    ValueType *__restrict__ omega,
+    const stopping_status *__restrict__ stop_status)
+{
+    const auto global_id = thread::get_thread_id_flat();
+
+    if (global_id >= nrhs) {
+        return;
+    }
+
+    if (!stop_status[global_id].has_stopped()) {
+        auto thr = omega[global_id];
+        omega[global_id] /= tht[global_id];
+        auto absrho =
+            abs(thr / (sqrt(real(tht[global_id])) * residual_norm[global_id]));
```

Comment:
does tht always be real? or always ignore the complex part?

Reply:
tht will always be real, but can be stored in a complex value type. It's the same as norm(t) squared, so since the norm is also needed, doing this saves one time computing th * t